packages/java-edition/src/binder/index.ts

```diff
@@ -10,130 +10,132 @@ import type {
 import { fileUtil, TaggableResourceLocationCategories } from '@spyglassmc/core'
 import { ReleaseVersion } from '../dependency/index.js'
 
+type ScopedVersion = { since?: ReleaseVersion; until?: ReleaseVersion }
+
+function JSONResource(
+	pack: 'datapack' | 'java_resource_pack',
+	dispatcher: FileCategory | TagFileCategory,
+	pathOrVersion: string | ScopedVersion,
+	version?: ScopedVersion,
+) {
+	let _version: ScopedVersion = {}
+
+	if (typeof pathOrVersion !== 'string') {
+		_version = pathOrVersion
+	} else if (version) {
+		_version = version
+	}
+	return [
+		typeof pathOrVersion === 'string' ? pathOrVersion : dispatcher,
+		{
+			pack,
+			category: dispatcher,
+			extname: '.json',
+			..._version,
+		} as const,
+	] as readonly [FileCategory | TagFileCategory, CategoryDef]
+}
+
+function DatapackJSON(
+	dispatcher: FileCategory | TagFileCategory,
+	pathOrVersion: string | ScopedVersion,
+	version?: ScopedVersion,
+) {
+	return JSONResource('datapack', dispatcher, pathOrVersion, version)
+}
+
+// function ResourcePackJSON(
+// 	dispatcher: FileCategory | TagFileCategory,
+// 	pathOrVersion: string | ScopedVersion,
+// 	version?: ScopedVersion,
+// ) {
+// 	return JSONResource('java_resource_pack', dispatcher, pathOrVersion, version)
+// }
+
+type CategoryDef = {
+	category: FileCategory | TagFileCategory
+	extname: string
+	pack: 'datapack' | 'java_resource_pack'
+} & ScopedVersion
```

Comment:
i think this is _slightly_ easier to read

```suggestion
interface CategoryDef extends ScopedVersion = {
	category: FileCategory | TagFileCategory
	extname: string
	pack: 'datapack' | 'java_resource_pack'
}
```